docs/data/data-grid/server-side-data/NestedPaginationGroupingCell.js

```diff
@@ -0,0 +1,155 @@
+import * as React from 'react';
+import composeClasses from '@mui/utils/composeClasses';
+import Box from '@mui/material/Box';
+import {
+  getDataGridUtilityClass,
+  useGridSelector,
+  useGridRootProps,
+  gridDataSourceLoadingIdSelector,
+  gridDataSourceErrorSelector,
+  useGridApiContext,
+  gridRowsLookupSelector,
+} from '@mui/x-data-grid-pro';
+import {
+  useGridPrivateApiContext,
+  createSelector,
+} from '@mui/x-data-grid-pro/internals';
+import useEventCallback from '@mui/utils/useEventCallback';
+import CircularProgress from '@mui/material/CircularProgress';
+
+const gridRowSelector = createSelector(
+  gridRowsLookupSelector,
+  (lookup, id) => lookup[id],
+);
+
+const useUtilityClasses = (ownerState) => {
+  const { classes } = ownerState;
+
+  const slots = {
+    root: ['treeDataGroupingCell'],
+    toggle: ['treeDataGroupingCellToggle'],
+    loadingContainer: ['treeDataGroupingCellLoadingContainer'],
+  };
+
+  return composeClasses(slots, getDataGridUtilityClass, classes);
+};
+
+function GridTreeDataGroupingCellIcon(props) {
+  const apiRef = useGridApiContext();
+  const rootProps = useGridRootProps();
+  const classes = useUtilityClasses({ classes: rootProps.classes });
+  const { rowNode, id, field, descendantCount, row, nestedLevelRef } = props;
+
+  const isDataLoading = useGridSelector(apiRef, gridDataSourceLoadingIdSelector, id);
+  const error = useGridSelector(apiRef, gridDataSourceErrorSelector, id);
+
+  const expanded = rowNode.childrenExpanded || row.expanded;
+
+  const handleClick = useEventCallback((event) => {
+    apiRef.current?.setRows([]);
+    if (!expanded) {
+      props.setExpandedRows((prev) => [
+        ...prev,
+        {
+          ...row,
+          groupingKey: rowNode.groupingKey,
+          expanded: true,
+          depth: nestedLevelRef.current,
+        },
+      ]);
+    } else if (row.expanded) {
+      props.setExpandedRows((prev) => {
+        const index = prev.findIndex((r) => r.id === id);
+        return prev.slice(0, index);
+      });
+    } else {
+      apiRef.current.setRowChildrenExpansion(id, !expanded);
+    }
+    apiRef.current.setCellFocus(id, field);
+    event.stopPropagation(); // TODO remove event.stopPropagation
```

Comment:
Does this need addressing before merge?

Reply:
I removed the comment and kept the `stopPropagation()` for now to avoid hitting `rowClick` event on children expansion.